src/vsr/superblock_free_set_encoded.zig

```diff
@@ -159,61 +157,69 @@ pub fn FreeSetEncodedType(comptime Storage: type) type {
                 .block_addresses = set.block_addresses,
                 .block_checksums = set.block_checksums,
             };
         }
 
         // These data are stored in the superblock header.
-        pub fn checkpoint_reference(set: *const Self) FreeSetReference {
+        pub fn checkpoint_reference(set: *const Self) vsr.SuperBlockFreeSetReference {
             assert(set.size == set.size_transferred);
             assert(set.callback == .none);
-            // TODO: Uncomment the two assertions once free set is no longer a part of superblock.
-            // These asserts are currently triggered by the superblock fuzzer which can't model
-            // free set properly without the grid.
-            // assert(set.grid.?.superblock.free_set.count_released() == set.block_count);
+            assert(set.grid.?.free_set.count_released() == set.block_count);
+
+            var storage_size = vsr.superblock.data_file_size_min;
+            if (set.grid.?.free_set.highest_address_acquired()) |address| {
+                assert(address > 0);
+                assert(set.size > 0);
+                storage_size += address * constants.block_size;
+            } else {
+                assert(set.size == 0);
+                assert(set.grid.?.free_set.count_released() == 0);
+            }
 
             if (set.block_count == 0) {
                 assert(set.size == 0);
-                // assert(set.grid.?.superblock.free_set.count_acquired() == 0);
+                assert(set.grid.?.free_set.count_acquired() == 0);
                 return .{
                     .checksum = set.checksum,
                     .last_block_address = 0,
                     .last_block_checksum = 0,
-                    .size = 0,
+                    .free_set_size = 0,
+                    .storage_size = storage_size,
                 };
             } else {
                 assert(set.size > 0);
                 return .{
                     .checksum = set.checksum,
                     .last_block_address = set.block_addresses[set.block_count - 1],
                     .last_block_checksum = set.block_checksums[set.block_count - 1],
-                    .size = set.size,
+                    .free_set_size = set.size,
+                    .storage_size = storage_size,
                 };
             }
         }
 
         pub fn open(
             set: *Self,
             grid: *Grid,
-            reference: FreeSetReference,
+            reference: vsr.SuperBlockFreeSetReference,
             callback: *const fn (set: *Self) void,
         ) void {
             set.grid = grid;
-            assert(!set.grid.?.superblock.free_set.opened);
+            assert(!set.grid.?.free_set.opened);
 
             assert(set.callback == .none);
             defer assert(set.callback == .open);
 
-            assert(reference.size % @sizeOf(FreeSet.Word) == 0);
+            assert(reference.free_set_size % @sizeOf(FreeSet.Word) == 0);
```

Comment:
We don't verify `reference.storage_size` at all during the `open` process, but the caller (replica) verifies it after the fact by checking that the current `.checkpoint_reference()` is meta.eql to the starting point from the manifest.